Intel SGX SDK Porting Guide (Linux) (#3542)
openenclave/openenclave: docs/GettingStartedDocs/IntelSDKPortingGuide.md

> # Intel SGX SDK to OpenEnclave SDK Porting Guide

Reviewer: ```suggestion
# Intel SGX SDK to Open Enclave SDK Porting Guide
```

> # Intel SGX SDK to OpenEnclave SDK Porting Guide

Reviewer: There are two separate Intel SGX SDKs... "SGX SDK for Linux" and "Intel SGX SDK for Windows".  We should avoid the unqualified term "Intel SGX SDK" throughout as it is a source of confusion.

I would like porting guide(s) for both.  Is the intent to put both into this document or have two documents?
if the latter, please change the title of the doc. 

And independent of the doc question, is this PR intending to cover both Windows and Linux?  If not, please change the title of the PR.

Author: The ultimate goal will be porting guides for both platforms but this PR and this doc is focus on Linux. I'll update the titles of both doc and PR.

Reviewer: Great, thanks.  We should file a separate github issue to track the porting guide for Windows then.

> Intel SGX SDK and OpenEnclave SDK share many design principles but differ in implementations. It requires source code changes to build enclaves with OpenEnclave SDK that were developed initially with Intel SGX SDK.

Reviewer: "Open Enclave" please :)

> Intel SGX SDK and OpenEnclave SDK share many design principles but differ in implementations. It requires source code changes to build enclaves with OpenEnclave SDK that were developed initially with Intel SGX SDK.

Author: I'll use `Open Enclave` throughout the doc. Thanks guys!

## Create Project Build Files

> OpenEnclave SDK uses `cmake` as the build system, so the first step of building an enclave is to create a `CMakeLists.txt` file for it. For those not familiar with `cmake`, a tutorial is available at https://cmake.org/cmake/help/latest/guide/tutorial/index.html.

Reviewer: This statement is not true for Windows.  Don't confuse the build system for the SDK itself, with the build system for apps.
For enclaves built using the SDK, there are multiple build systems supported.  For example, for Visual Studio as a build system, see https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/VisualStudioWindows.md

```cmake
find_package(OpenEnclave CONFIG REQUIRED)
```

Now generate the ECall/OCall bridge/proxy routines for the enclave.

```cmake
add_custom_command(
  OUTPUT MyEnclave_t.h MyEnclave_t.c MyEnclave_args.h
  DEPENDS ${CMAKE_SOURCE_DIR}/MyEnclave.edl
  COMMAND openenclave::oeedger8r --trusted ${CMAKE_SOURCE_DIR}/MyEnclave.edl
          --search-path ${OE_INCLUDEDIR})
```

Finally, build the enclave as an executable. Please don't forget to include `MyEnclave_t.c` (generated above) in the source list.

```cmake
add_executable(MyEnclave MyEnclave.cpp ${CMAKE_CURRENT_BINARY_DIR}/MyEnclave_t.c)

# Current API version
target_compile_definitions(MyEnclave PUBLIC OE_API_VERSION=2)

# Needed by the generated MyEnclave_t.c
target_include_directories(MyEnclave PRIVATE ${CMAKE_CURRENT_BINARY_DIR})

target_link_libraries(MyEnclave openenclave::oeenclave openenclave::oelibc)
```

> Enclaves must be signed before they can be loaded. How to sign an enclave depends on how the signing private key is managed in your project. For demonstration purpose, the following snippet generates a random RSA key pair to sign the enclave. An enclave configuration file (`MyEnclave.conf` below) may also be provided. Details of enclave configuration/settings are described later in this document.

Reviewer: ```suggestion
Enclaves must be signed before they can be loaded. How to sign an enclave depends on how the signing private key is managed in your project. For demonstration purposes, the following snippet generates a random RSA key pair to sign the enclave. An enclave configuration file (`MyEnclave.conf` below) may also be provided. Details of enclave configuration/settings are described later in this document.
```

```cmake
# Generate key
add_custom_command(
  OUTPUT private.pem public.pem
  COMMAND openssl genrsa -out private.pem -3 3072
  COMMAND openssl rsa -in private.pem -pubout -out public.pem)

# Sign enclave
add_custom_command(
  OUTPUT MyEnclave.signed
  DEPENDS MyEnclave MyEnclave.conf private.pem
  COMMAND openenclave::oesign sign -e $<TARGET_FILE:MyEnclave>
          -c ${CMAKE_SOURCE_DIR}/MyEnclave.conf -k private.pem)
```

> Audiences are encouraged to look for complete examples under */path/to/OpenEnclave_InstallDir*/share/openenclave/samples directory.

Reviewer 2: ```suggestion
Readers are encouraged to look for complete examples under */path/to/OpenEnclave_InstallDir*/share/openenclave/samples directory.
```

## Migrate Enclave Settings

> **Enclave Settings**, also known as **Enclave Metadata**, refer to information consumed by enclave loaders to instantiate enclaves, such as heap size, stack size, number of trusted hardware threads (i.e. number of TCS's), etc.

Reviewer: ```suggestion
**Enclave Settings**, also known as **Enclave Metadata**, refers to information consumed by enclave loaders to instantiate enclaves, such as heap size, stack size, number of trusted hardware threads (i.e., number of TCS's), etc.
```

Enclave settings are specified in human readable text formats by enclave developers, referred to as *configuration files*. Both Intel SGX SDK and OpenEnclave SDK provide tools to compile configuration files into their binary form and embed them into the final enclave image. However, they differ in both format and feature set.

### Configuration File Formats

> Intel SGX SDK adopted an XML format for encoding enclave settings in text form, which is usually named as *Enclave*.config.xml. The signing tool (i.e. `sgx_sign`) converts it into binary form and stores it in a dedicated section (i.e. `.sgxmeta`) of the enclave's ELF image before it calculates the enclave's measurement (i.e. `SIGSTRUCT::MRENCLAVE`). At runtime, the ELF section `.sgxmeta` is consumed by the enclave loader to instatiate the exact enclave that matches the measurement caculated by the signing tool.

Reviewer: ```suggestion
The Intel SGX SDK adopted an XML format for encoding enclave settings in text form, which is usually named as *Enclave*.config.xml. The signing tool (i.e., `sgx_sign`) converts it into binary form and stores it in a dedicated section (i.e. `.sgxmeta`) of the enclave's ELF image before it calculates the enclave's measurement (i.e., `SIGSTRUCT::MRENCLAVE`). At runtime, the ELF section `.sgxmeta` is consumed by the enclave loader to instantiate the exact enclave that matches the measurement calculated by the signing tool.
```
Also please split into multiple source lines, as you can see that github has a hard time correctly showing the diffs with such a long line, which makes it harder to review.

Below comes from the sample code - SampleEnclave, of Intel SGX SDK.
```xml
<EnclaveConfiguration>
  <ProdID>0</ProdID>
  <ISVSVN>0</ISVSVN>
  <StackMaxSize>0x40000</StackMaxSize>
  <HeapMaxSize>0x100000</HeapMaxSize>
  <TCSNum>10</TCSNum>
  <TCSPolicy>1</TCSPolicy>
  <!-- Recommend changing 'DisableDebug' to 1 to make the enclave undebuggable for enclave release -->
  <DisableDebug>0</DisableDebug>
  <MiscSelect>0</MiscSelect>
  <MiscMask>0xFFFFFFFF</MiscMask>
</EnclaveConfiguration>
```

> Rather than XML, OpenEnclave SDK uses plaintext files instead. A configuration file, usually named as *enclave*.conf, is supplied to the signing tool (i.e. `oesign`) command line to govern the instantiation of the enclave. The compiled metadata is then stored in a dedicated ELF section named `.oeinfo`. Additionally, OpenEnclave SDK provides `OE_SET_ENCLAVE_SGX`, a C macro for embedding default enclave settings in C source files. *enclave*.conf is in fact optional and is necessary only if some of those defaults provided to `OE_SET_ENCLAVE_SGX` macro need overridden. Under the hood, `OE_SET_ENCLAVE_SGX` is expanded to instantiation of an `oe_sgx_enclave_properties_t` structure in `.oeinfo` section. Detailed information can be found in [buildandsign.md](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/buildandsign.md) in OpenEnclave SDK documentation.

Reviewer: ```suggestion
Rather than XML, OpenEnclave SDK uses plaintext files instead. A configuration file, usually named as *enclave*.conf, is supplied to the signing tool (i.e., `oesign`) command line to govern the instantiation of the enclave. The compiled metadata is then stored in a dedicated ELF section named `.oeinfo`. Additionally, OpenEnclave SDK provides `OE_SET_ENCLAVE_SGX`, a C macro for embedding default enclave settings in C source files. *enclave*.conf is in fact optional and is necessary only if some of those defaults provided to `OE_SET_ENCLAVE_SGX` macro need overridden. Under the hood, `OE_SET_ENCLAVE_SGX` is expanded to instantiation of an `oe_sgx_enclave_properties_t` structure in `.oeinfo` section. Detailed information can be found in [buildandsign.md](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/buildandsign.md) in OpenEnclave SDK documentation.
```
Also please split into multiple source lines so you don't have to visually play "spot the differences" in suggestions like above.

|`<ReservedMemMinSize>`||`uint64_t`
|`<ReservedMemInitSize>`||`uint64_t`
|`<ReservedMemExecutable>`||`uint64_t`
|`<MiscSelect>`||`uint32_t`|`SIGSTRUCT::MISCSELECT` - selects extended information to be reported on AEX|OpenEnclave does *NOT* support configuring this field currently but hard-codes it to `0` (i.e. no MISC features are enabled).
|`<MiscMask>`||`uint32_t`|`SIGSTRUCT::MISCMASK` - selects `MISCSELECT` bits to enforce|OpenEnclave does *NOT* support configuring this field currently but hard-codes it to `0xffffffff` (i.e. all bits are enforced).
|`<EnableKSS>`||`bool`|`1` to enable *Key Separation and Sharing*|OpenEnclave supports only SGX v1 at the moment.
|`<ISVFAMILYID_H>`||`uint64_t`|This, along with `<ISVFAMILYID_L>` below, forms 16-byte `SIGSTRUCT::ISVFAMILYID`|OpenEnclave supports only SGX v1 at the moment.
|`<ISVFAMILYID_L>`||`uint64_t`|See above|
|`<ISVEXTPRODID_H>`||`uint64_t`|This, along with `<ISVEXTPRODID_L>` below, forms 16-byte `SIGSTRUCT::ISVEXTPRODID`|OpenEnclave supports only SGX v1 at the moment.
|`<ISVEXTPRODID_L>`||`uint64_t`|See above|

> As mentioned in the *Notes* column above, certain missing feature, such as those controlling SGX enclave attribute bits (e.g. `<LaunchKey>`, `<ProvisionKey>`), could still be enabled by setting `oe_sgx_enclave_properties_t::config.attributes` manually, even though they aren't supported explicitly by the `OE_SET_ENCLAVE_SGX` macro or *enclave*.conf file. Below is an example that grants the enclave access to `PROVISION_KEY`.

Reviewer: ```suggestion
As mentioned in the *Notes* column above, certain missing feature, such as those controlling SGX enclave attribute bits (e.g., `<LaunchKey>`, `<ProvisionKey>`), could still be enabled by setting `oe_sgx_enclave_properties_t::config.attributes` manually, even though they aren't supported explicitly by the `OE_SET_ENCLAVE_SGX` macro or *enclave*.conf file. Below is an example that grants the enclave access to `PROVISION_KEY`.
```

```C
OE_INFO_SECTION_BEGIN
volatile const oe_sgx_enclave_properties_t oe_enclave_properties_sgx = {
    .header = {.size = sizeof(oe_sgx_enclave_properties_t),
               .enclave_type = OE_ENCLAVE_TYPE_SGX,
               .size_settings = {.num_heap_pages = 512,
                                 .num_stack_pages = 512,
                                 .num_tcs = 4}},
    .config = {.product_id = 1234,
               .security_version = 5678,
               .attributes = OE_SGX_FLAGS_PROVISION_KEY |
                             OE_MAKE_ATTRIBUTES(0)},
    .end_marker = 0xecececececececec};
OE_INFO_SECTION_END
```

If compatibilities with both SDKs are desired, avoid using features specific to either SDK.

> As a final note, neither Intel SGX SDK nor OpenEnclave SDK provides configuration settings for enabling/disabling X features (e.g. AVX, AVX-512, etc.) explicitly. OpenEnclave's SGX enclave loader uses the enabled X features on the local platform to initialize `SECS::ATTRIBUTES::XFRM`, and hard-codes `SIGSTRUCT::ATTRIBUTEMASK::XFRM` to `0`. That is, X features are *NOT* enforced and must *NOT* be relied upon for security.

Reviewer: ```suggestion
As a final note, neither Intel SGX SDK nor OpenEnclave SDK provides configuration settings for enabling/disabling X features (e.g., AVX, AVX-512, etc.) explicitly. OpenEnclave's SGX enclave loader uses the enabled X features on the local platform to initialize `SECS::ATTRIBUTES::XFRM`, and hard-codes `SIGSTRUCT::ATTRIBUTEMASK::XFRM` to `0`. That is, X features are *NOT* enforced and must *NOT* be relied upon for security.
```

## Migrate ECall/OCall Definitions (EDL Files)

Both Intel SGX SDK and OpenEnclave SDK support the same grammer for defining trusted/untrusted functions (aka. ECalls/OCalls) in EDL files. However, built-in OCalls are defined in different headers. Intel SGX SDK's built-in OCalls are defined in `sgx_tstdc.edl` while OpenEnclave SDK's are defined in `platform.edl`.

> Most EDL files include (by `include` statements) common C headers for both host and enclave sides. The most commonly included header is the one defining SGX architectural structures, which is `arch.h` in Intel SGX SDK or `openenclave/bits/sgx/sgxtypes.h` in OpenEnclave SDK. Please also note that some structures may be named differently, e.g. the EINITTOKEN architectural structure is defined as `token_t` in Intel SGX SDK but `einittoken_t` in OpenEnclave SDK.

Reviewer: ```suggestion
Most EDL files include (by `include` statements) common C headers for both host and enclave sides. The most commonly included header is the one defining SGX architectural structures, which is `arch.h` in Intel SGX SDK or `openenclave/bits/sgx/sgxtypes.h` in OpenEnclave SDK. Please also note that some structures may be named differently, e.g., the EINITTOKEN architectural structure is defined as `token_t` in Intel SGX SDK but `einittoken_t` in OpenEnclave SDK.
```

The code snippet below shows a way to include/import C headers and EDL definitions conditionally, in order to be compatible with both Intel SGX and OpenEnclave SDKs.

```
enclave {
#ifdef OEEDGER8R
    include "openenclave/bits/sgx/sgxtypes.h";
    from "openenclave/edl/sgx/platform.edl" import *;
#else
    include "arch.h";
    from "sgx_tstdc.edl" import *;
#endif
    /* ECall/OCall definitions go here */
}
```

The macro `OEEDGER8R` used above could be defined by appending `-DOEEDGER8R` to `oeedger8r` command line. Please note that only `oeedger8r` supports defining macros at command line, even though both `sgx_edger8r` and `oeedger8r` support preprocessing.

> The last thing worth noting is that OpenEnclave doesn't support nested ECall (i.e. an ECall in the context of an OCall) like Intel SGX SDK does. Existing enclaves making use of nested ECalls need to be reworked to be compatible with OpenEnclave SDK.

Reviewer: ```suggestion
The last thing worth noting is that OpenEnclave doesn't support nested ECall (i.e., an ECall in the context of an OCall) like the Intel SGX SDK does. Existing enclaves making use of nested ECalls need to be reworked to be compatible with the Open Enclave SDK.
```

## Port C/C++ Source Code

Given similarities in the architectures of both SDKs, there shall not be any significant code flow/logic changes required. However, source code incompatibilities still exist in
* Header files - They are structured and/or named differently. Fortunately, OpenEnclave provides 2 comprehensive headers, namely `openenclave/enclave.h` and `openenclave/host.h`, to be included by trusted and untrusted code, respectively. A single `#include` should suffice in most cases.
* APIs - Most OpenEnclave APIs are prefixed by `oe_` while Intel's APIs are by `sgx_`. Moreover, some APIs may take parameters in different orders.
* Structure definitions - Structure members may be named differently. Some structures are organized differently. For example, Intel SGX SDK defines EINITTOKEN as `token_t` with all MAC'ed fields captured in a child structure `launch_body_t`; while OpenEnclave defines it as a flat `einittoken_t` structure.
* Crypto lib - Intel SGX SDK supports 2 crypto libs - IPP and OpenSSL, and provides a wrapper layer to unify crypto APIs. OpenEnclave supports only MbedTLS and provides no wrapper.

> Generally, there's no good way to fast port source code rather than reacting to compiler errors and substituting text strings manually.

Author: Is there a template for the "Authors" section or is there a style that I can follow?

> Generally, there's no good way to fast port source code rather than reacting to compiler errors and substituting text strings manually.

Reviewer 2: You can find the Authors section in other design docs, e.g. [AttestationApiRequirements.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/AttestationApiRequirements.md#authors), [SGX_QuoteEx_Integration.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/SGX_QuoteEx_Integration.md#authors), and [ExtendedEnclaveInitializationData.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/ExtendedEnclaveInitializationData.md#authors)
